browser/search_engines/search_engine_provider_service.cc

```diff
@@ -76,3 +85,47 @@ void SearchEngineProviderService::ChangeToNormalWindowSearchEngineProvider() {
   otr_template_url_service_->SetUserSelectedDefaultSearchProvider(
       &normal_url);
 }
+
+void SearchEngineProviderService::UseExtensionSearchProvider() {
+#if BUILDFLAG(ENABLE_EXTENSIONS)
+  DCHECK(ShouldUseExtensionSearchProvider());
+
+  const auto* extension_provider_url =
+      original_template_url_service_->GetDefaultSearchProvider();
+  auto data = extension_provider_url->data();
+  data.id = kInvalidTemplateURLID;
+
+  // Can't add same turl again to service.
+  if (CouldAddExtensionTemplateURL(extension_provider_url)) {
+    auto type = extension_provider_url->type();
+    auto extension_id = extension_provider_url->GetExtensionId();
+    extensions::ExtensionPrefs* prefs =
+        extensions::ExtensionPrefs::Get(otr_profile_->GetOriginalProfile());
+    auto time = prefs->GetInstallTime(extension_id);
+
+    auto turl =
+        std::make_unique<TemplateURL>(data, type, extension_id, time, true);
+
+    otr_template_url_service_->Add(std::move(turl));
+  }
+
+  otr_profile_->GetPrefs()->Set(
+      DefaultSearchManager::kDefaultSearchProviderDataPrefName,
+      *TemplateURLDataToDictionary(data));
+#endif
+}
+
+bool SearchEngineProviderService::ShouldUseExtensionSearchProvider() {
+  return original_template_url_service_->IsExtensionControlledDefaultSearch();
+}
+
+bool SearchEngineProviderService::CouldAddExtensionTemplateURL(
+    const TemplateURL* url) {
+  DCHECK_NE(TemplateURL::NORMAL, url->type());
+  for (const auto* turl : otr_template_url_service_->GetTemplateURLs()) {
+    if (url->type() == turl->type() &&
```

Comment:
`DCHECK(url)` and `DCHECK(turl)` would be nice.

Reply:
done.